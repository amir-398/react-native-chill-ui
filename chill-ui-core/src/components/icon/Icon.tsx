import { cn } from '@utils';
import { useState } from 'react';
import { tv } from 'tailwind-variants';
import { Pressable, TouchableOpacity } from 'react-native';

import type { IconProps } from '../../types/icon.types';

import styles from './Icon.style';
import CustomIcon from './CustomIcon';
import { RipplePressable } from '../ripplePressable';
import { isNativeWindInstalled } from '../../utils/hybrid/nativewindDetector';
import { classNamePropsHandler } from '../../utils/hybrid/classNamePropsHandler';

// padding  variant
export const paddingVr = tv({
  base: 'p-0',
  variants: {
    size: {
      '2xl': 'p-5',
      '2xs': 'p-[1px]',
      '3xl': 'p-6',
      lg: 'p-3',
      md: 'p-2',
      sm: 'p-1.5',
      xl: 'p-4',
      xs: 'p-0.5',
    },
  },
});
export const IconSizeVr = tv({
  base: 'size-6',
  variants: {
    size: {
      '2xl': 'size-9',
      '2xs': 'size-3',
      '3xl': 'size-10',
      lg: 'size-7',
      md: 'size-6',
      sm: 'size-5',
      xl: 'size-8',
      xs: 'size-4',
    },
  },
});

/**
 * Icon component that displays SVG icons with customizable size, color, and press interactions.
 * Supports both static and interactive icons with press effects.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * // With NativeWind - Basic icon
 * <Icon name="bell-solid" size="md" color="#000" />
 *
 * // With NativeWind - Clickable icon with press effect
 * <Icon
 *   name="star-solid"
 *   size="lg"
 *   color="#FFD700"
 *   onPress={() => console.log('Icon pressed')}
 *   hasPressEffect={true}
 * />
 *
 * // With NativeWind - Custom press effect
 * <Icon
 *   name="heart-solid"
 *   size="md"
 *   color="#FF6B6B"
 *   onPress={handleLike}
 *   pressEffectClassName="bg-red-100"
 * />
 *
 * // Without NativeWind (fallback) - Basic icon
 * <Icon name="bell-solid" size="md" color="#000" />
 *
 * // Without NativeWind (fallback) - Clickable icon
 * <Icon
 *   name="star-solid"
 *   size="lg"
 *   color="#FFD700"
 *   onPress={() => console.log('Icon pressed')}
 *   hasPressEffect={true}
 *   style={{ backgroundColor: '#f3f4f6' }}
 * />
 * ```
 *
 * @param name - Icon name from the available icon set
 * @param size - Icon size variant (default: 'md')
 * @param color - Icon color (default: '#000')
 * @param onPress - Callback function when icon is pressed
 * @param hasPressEffect - Whether to show press effect (default: true)
 * @param pressEffectClassName - Custom CSS classes for press effect (NativeWind only)
 * @param pressEffectSize - Size of press effect padding
 * @param className - Custom CSS classes for the icon (NativeWind only)
 * @param style - Inline styles for fallback without NativeWind
 */
export default function Icon(props: IconProps) {
  classNamePropsHandler(props, 'Icon');
  const {
    as,
    className,
    color = '#000',
    hasPressEffect = true,
    name,
    onPress,
    pressEffectClassName = 'bg-dark/10',
    pressEffectSize,
    size = 'md',
    style,
    ...rest
  } = props;
  const [isPressed, setIsPressed] = useState(false);

  if (!onPress) {
    if (isNativeWindInstalled()) {
      return <CustomIcon name={name} className={cn(IconSizeVr({ size }), className)} color={color} />;
    }

    const iconSizeStyle = styles[`size${size.charAt(0).toUpperCase() + size.slice(1)}` as keyof typeof styles];
    return <CustomIcon name={name} style={[styles.base, iconSizeStyle, style]} color={color} />;
  }

  if (isNativeWindInstalled()) {
    return (
      <Pressable
        // @ts-ignore
        className={cn(
          'rounded-full',
          hasPressEffect && paddingVr({ size: pressEffectSize || size }),
          isPressed ? pressEffectClassName : 'bg-transparent',
        )}
        onPressIn={() => setIsPressed(true)}
        onPressOut={() => setIsPressed(false)}
        onPress={onPress}
        {...rest}
      >
        <CustomIcon name={name} className={cn(IconSizeVr({ size }), className)} style={style} color={color} />
      </Pressable>
    );
  }

  const iconSizeStyle = styles[`size${size.charAt(0).toUpperCase() + size.slice(1)}` as keyof typeof styles];
  const paddingStyle = hasPressEffect
    ? styles[
        `padding${(pressEffectSize || size).charAt(0).toUpperCase() + (pressEffectSize || size).slice(1)}` as keyof typeof styles
      ]
    : null;

  const pressEffectStyle = ({ pressed }: { pressed: boolean }) =>
    pressed && hasPressEffect ? styles.pressEffectDefault : styles.transparent;

  switch (as) {
    case 'Pressable':
      return (
        <Pressable
          style={({ pressed }) => [
            styles.pressableBase,
            styles.pressableContainer,
            paddingStyle,
            pressEffectStyle({ pressed }),
            style,
          ]}
          onPress={onPress}
          {...rest}
        >
          <CustomIcon name={name} style={[styles.base, iconSizeStyle]} color={color} />
        </Pressable>
      );

    case 'RipplePressable':
      return (
        <RipplePressable onPress={onPress} {...rest}>
          <CustomIcon name={name} style={[styles.base, iconSizeStyle]} color={color} />
        </RipplePressable>
      );

    default:
      return (
        <TouchableOpacity onPress={onPress} style={styles.pressableBase} {...rest}>
          <CustomIcon name={name} style={[styles.base, iconSizeStyle]} color={color} />
        </TouchableOpacity>
      );
  }
}
