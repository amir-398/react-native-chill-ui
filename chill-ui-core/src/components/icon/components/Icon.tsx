import { IconProps } from '@types';
import { Pressable, TouchableOpacity } from 'react-native';
import { RipplePressable } from '@components/ripplePressable';
import { cn, classNamePropsHandler, classNameHandler, styleHandler } from '@utils';

import CustomIcon from './CustomIcon';
import { iconDefaultProps } from '../utils/defaultProps';
import { iconSizeSv, iconPaddingSv, iconPressEffectSv, styles } from '../styles/Icon.styles';
import { iconSizeTv, iconPaddingTv, iconPressableTv, iconClassname } from '../styles/Icon.variants';

/**
 * Icon component that displays SVG icons with customizable size, color, and press interactions.
 * Supports both static and interactive icons with press effects.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * // Basic icon
 * <Icon name="heart-solid" />
 *
 * // Customized with size and color
 * <Icon name="star-solid" size="lg" color="#FFD700" />
 * ```
 * @param name - Icon name from the available icon set (required)
 * @param size - Icon size variant: '2xs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' (default: 'md')
 * @param color - Icon color (default: '#000')
 * @param onPress - Callback function when icon is pressed
 * @param hasPressEffect - Whether to show press effect when pressed (default: true)
 * @param pressEffectSize - Size of the press effect padding
 * @param pressEffectStyle - Custom styles for the press effect
 * @param as - Component to use when pressable: 'pressable' | 'touchable-opacity' | 'ripple-pressable' (default: 'pressable')
 * @param className - Custom CSS classes (used with NativeWind)
 * @param style - Additional inline styles
 */
export default function Icon(props: IconProps) {
  classNamePropsHandler(props, 'Icon');

  const {
    as = iconDefaultProps.as,
    className,
    color = iconDefaultProps.color,
    hasPressEffect = iconDefaultProps.hasPressEffect,
    name,
    onPress,
    pressEffectSize,
    pressEffectStyle,
    size = iconDefaultProps.size,
    style,
    ...rest
  } = props;

  const iconStyle = iconSizeSv({ size });

  const iconContent = (
    <CustomIcon
      name={name}
      {...classNameHandler(cn(iconSizeTv({ size }), { [cn(iconClassname, className)]: !onPress }))}
      {...styleHandler({ defaultStyle: [iconStyle, !onPress && styles.iconBase], style: !onPress ? style : undefined })}
      color={color}
    />
  );

  if (!onPress) {
    return iconContent;
  }

  switch (as) {
    case 'ripple-pressable':
      return (
        <RipplePressable
          {...classNameHandler(cn(iconClassname, className))}
          {...styleHandler({ style })}
          onPress={onPress}
          {...rest}
        >
          {iconContent}
        </RipplePressable>
      );
    case 'touchable-opacity':
      return (
        <TouchableOpacity
          {...classNameHandler(cn(iconClassname, className))}
          {...styleHandler({ style })}
          onPress={onPress}
          {...rest}
        >
          {iconContent}
        </TouchableOpacity>
      );

    case 'pressable':
    default:
      return (
        <Pressable
          {...classNameHandler(
            cn(
              iconClassname,
              iconPressableTv({ hasPressEffect }),
              hasPressEffect && iconPaddingTv({ size: pressEffectSize ?? size }),
              className,
            ),
          )}
          style={({ pressed }) =>
            styleHandler({
              defaultStyle: [
                styles.iconBase,
                hasPressEffect && iconPaddingSv({ pressEffectSize: pressEffectSize ?? size }),
                iconPressEffectSv({ pressed: hasPressEffect && pressed }),
              ],
              style: [pressed && pressEffectStyle, style],
            }).style
          }
          onPress={onPress}
          {...rest}
        >
          {iconContent}
        </Pressable>
      );
  }
}
