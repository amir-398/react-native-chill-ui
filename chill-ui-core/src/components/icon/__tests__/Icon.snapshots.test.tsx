import { render } from '@testing-library/react-native';

import Icon from '../components/Icon';

// Mock du composant CustomIcon
jest.mock('../components/CustomIcon', () => {
  return function MockCustomIcon({ name, color, style }: any) {
    return (
      <div testID={`icon-${name}`} style={style} data-color={color}>
        {name}
      </div>
    );
  };
});

// Mock du composant RipplePressable
jest.mock('../../ripplePressable', () => ({
  RipplePressable: ({ children, onPress, style, className, ...props }: any) => (
    <div testID="ripple-pressable" onPress={onPress} style={style} className={className} {...props}>
      {children}
    </div>
  ),
}));

describe('Icon Component Snapshots', () => {
  it('should match snapshot for basic icon', () => {
    const tree = render(<Icon name="heart-solid" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for icon with size', () => {
    const tree = render(<Icon name="star-solid" size="lg" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for pressable icon', () => {
    const tree = render(<Icon name="heart-solid" onPress={() => {}} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for icon with custom color', () => {
    const tree = render(<Icon name="star-solid" color="#FF6B6B" />);
    expect(tree).toMatchSnapshot();
  });
});
