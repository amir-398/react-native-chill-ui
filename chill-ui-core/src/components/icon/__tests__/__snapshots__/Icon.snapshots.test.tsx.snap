// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component Snapshots should match snapshot for basic icon 1`] = `
<div
  data-color="#000"
  testID="icon-heart-solid"
>
  heart-solid
</div>
`;

exports[`Icon Component Snapshots should match snapshot for icon with custom color 1`] = `
<div
  data-color="#FF6B6B"
  testID="icon-star-solid"
>
  star-solid
</div>
`;

exports[`Icon Component Snapshots should match snapshot for icon with size 1`] = `
<div
  data-color="#000"
  testID="icon-star-solid"
>
  star-solid
</div>
`;

exports[`Icon Component Snapshots should match snapshot for pressable icon 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="self-start rounded-full active:bg-[rgba(0,0,0,0.1)] p-[8px]"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        false,
        undefined,
      ]
    }
  >
    <div
      data-color="#000"
      testID="icon-heart-solid"
    >
      heart-solid
    </div>
  </View>
</View>
`;
